src/Features/Core/Portable/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+        /// is {Derived}
+        /// </summary>
+        /// <param name="diagnosticId">The Id of diagonostic</param>
+        /// <param name="spanNode">the innermost node that contains the span</param>
+        /// <param name="potentialConversionTypes"> Output (target expression, potential conversion type) pairs</param>
+        /// <returns>
+        /// True, if there is at least one potential conversion pair, and they are assigned to "potentialConversionTypes"
+        /// False, if there is no potential conversion pair.
+        /// </returns>
+        protected abstract bool TryGetTargetTypeInfo(Document document, SemanticModel semanticModel, SyntaxNode root,
+            string diagnosticId, TExpressionSyntax spanNode, CancellationToken cancellationToken,
+            out ImmutableArray<Tuple<TExpressionSyntax, ITypeSymbol>> potentialConversionTypes);
+
+        protected abstract bool IsObjectCreationExpression(TExpressionSyntax targetNode);
+
+        protected abstract bool IsConversionUserDefined(
+            SemanticModel semanticModel, TExpressionSyntax expression, ITypeSymbol type);
+
+        protected ImmutableArray<Tuple<TExpressionSyntax, ITypeSymbol>> FilterValidPotentialConversionTypes(
+            SemanticModel semanticModel,
+            ArrayBuilder<Tuple<TExpressionSyntax, ITypeSymbol>> mutablePotentialConversionTypes)
+        {
+            using var _ = ArrayBuilder<Tuple<TExpressionSyntax, ITypeSymbol>>.GetInstance(out var validPotentialConversionTypes);
+            foreach (var conversionTuple in mutablePotentialConversionTypes)
+            {
+                var targetNode = conversionTuple.Item1;
+                var targetNodeConversionType = conversionTuple.Item2;
+
+                // For cases like object creation expression. for example:
+                // Derived d = [||]new Base();
+                // It is always invalid except the target node has explicit conversion operator or is numeric.
+                if (IsObjectCreationExpression(targetNode)
+                    && !IsConversionUserDefined(semanticModel, targetNode, targetNodeConversionType))
+                {
+                    continue;
+                }
+
+                validPotentialConversionTypes.Add(conversionTuple);
+            }
+            return validPotentialConversionTypes.Distinct().ToImmutableArray();
+        }
+
+        protected abstract bool ClassifyConversionExists(
+            SemanticModel semanticModel, TExpressionSyntax expression, ITypeSymbol type);
+
+        protected abstract bool IsConversionIdentity(
+            SemanticModel semanticModel, TExpressionSyntax expression, ITypeSymbol type);
+
+        protected abstract SeparatedSyntaxList<SyntaxNode> GetArguments(SyntaxNode argumentList);
+
+        protected abstract SyntaxNode GenerateNewArgument(SyntaxNode oldArgument, ITypeSymbol conversionType);
+
+        protected abstract TExpressionSyntax? GetArgumentExpression(SyntaxNode argument);
```

Comment:
1. GetExpressionOfArgument.
2. Can/should be an ISyntaxFact.
3.  Odd that we have TExpressionSyntax but not TArgumentyntax.